openquake/engine/calculators/hazard/general.py

```diff
@@ -341,24 +341,26 @@ def initialize_realizations(self):
 
         rlzs_assoc = cm.get_rlzs_assoc()
         gsims_by_trt_id = rlzs_assoc.get_gsims_by_trt_id()
-        for rlz, gsim_by_trt in zip(
-                rlzs_assoc.realizations, rlzs_assoc.gsim_by_trt):
-            lt_model = models.LtSourceModel.objects.get(
-                hazard_calculation=self.job,
-                sm_lt_path=self.tilepath + rlz.sm_lt_path)
+        smodels = [sm for sm in self._source_models
+                   if sm.trtmodel_set.filter(num_ruptures__gt=0)]
+        for lt_model, rlzs in zip(smodels, rlzs_assoc.rlzs_by_smodel):
             trt_models = lt_model.trtmodel_set.filter(num_ruptures__gt=0)
```

Comment:
It is needed twice. A source model contains several trt models. A source model is discarded only if only of its trt models contains zero ruptures.